src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -292,7 +293,8 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 bx.load(addr, self.fn_ty.ret.layout.align.abi)
             }
         };
-        bx.ret(llval);
+        let llvalc = bx.flat_addr_cast(llval);
```

Comment:
IMO the only way this PR will be accepted without an RFC is if *everything* is cast to the flat address space *as soon as possible* (as opposed to when needed).